src/wp-admin/includes/plugin.php

```diff
@@ -1418,45 +1408,69 @@ function add_submenu_page( $parent_slug, $page_title, $menu_title, $capability,
 	}
 
 	$new_sub_menu = array( $menu_title, $capability, $menu_slug, $page_title );
-	if ( ! is_int( $position ) ) {
-		if ( null !== $position ) {
-			_doing_it_wrong(
-				__FUNCTION__,
-				sprintf(
-					/* translators: %s: add_submenu_page() */
-					__( 'The seventh parameter passed to %s should be an integer representing menu position.' ),
-					'<code>add_submenu_page()</code>'
-				),
-				'5.3.0'
-			);
-		}
 
+	if (
+		null === $position
+		// BC: `$position` is invalid.
+		|| ( ! is_string( $position ) && ! is_int( $position ) && ! is_float( $position ) )
+		// BC: The parent does not exist.
+		|| ! isset( $submenu[ $parent_slug ] )
+		// BC: `$position` is greater than or equal to the number of items.
+		|| (int) $position >= count( $submenu[ $parent_slug ] )
+	) {
 		$submenu[ $parent_slug ][] = $new_sub_menu;
 	} else {
-		// Append the submenu if the parent item is not present in the submenu,
-		// or if position is equal or higher than the number of items in the array.
-		if ( ! isset( $submenu[ $parent_slug ] ) || $position >= count( $submenu[ $parent_slug ] ) ) {
-			$submenu[ $parent_slug ][] = $new_sub_menu;
+		// BC: Test for a negative position.
+		if ( 0 > (int) $position ) {
+			$position = 0;
+		}
+
+		// BC: Handle a `$position` of `0`.
+		if ( 0 === $position ) {
+			/*
+			 * BC:
+			 *
+			 * == TODO ==
+			 *
+			 * Add $new_sub_menu to the start of $submenu[ $parent_slug ].
+			 *
+			 * Must preserve the keys in $submenu[ $parent_slug ].
+			 * For BC, should ensure the new item is first, regardless of sort.
+			 *
+			 * This current version increments all existing keys between 0 and 1.
+			 *
+			 * 1. Should this mechanism be kept for BC?
+			 * 2. Does this mechanism defeat the purpose of consistent ordering?
+			 *
+			 * Once these are answered, the code block below either needs
+			 * to removed or evaluated/improved.
+			 */
+			$current_first_key = array_key_first( $submenu[ $parent_slug ] );
```

Comment:
array_key_first() is PHP 7.3+

Reply:
@sybrew thanks fro looking at these. I meant to close this PR as superseded by the other one you've seen.